compiler/passes/src/common/graph/mod.rs

```diff
@@ -0,0 +1,205 @@
+// Copyright (C) 2019-2022 Aleo Systems Inc.
+// This file is part of the Leo library.
+
+// The Leo library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The Leo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
+
+use leo_span::Symbol;
+
+use indexmap::{IndexMap, IndexSet};
+use std::fmt::Debug;
+use std::hash::Hash;
+
+/// A node in a graph.
+pub trait Node: Copy + 'static + Eq + PartialEq + Debug + Hash {}
+
+impl Node for Symbol {}
+
+/// Errors in graph operations.
+#[derive(Debug)]
+pub enum GraphError<N: Node> {
+    /// An error that is emitted when a cycle is detected in the graph. Contains the path of cycle.
+    CycleDetected(Vec<N>),
+}
+
+/// A directed graph.
+#[derive(Debug)]
+pub struct DiGraph<N: Node> {
+    /// The set of nodes in the graph.
+    nodes: IndexSet<N>,
+    // TODO: Better name.
+    /// The directed edges in the graph.
```

Comment:
Here we could say that the finite map represents the edges from the domain to the range, so to speak. Which can be inferred from subsequent code, but it's good to specify, as one could represent them the other way.